net/multi_listen_darwin.go

```diff
@@ -0,0 +1,94 @@
+//go:build darwin
+// +build darwin
+
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package net
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"syscall"
+)
+
+// Accept is part of net.Listener interface.
+func (ml *multiListener) accept() (net.Conn, error) {
+	for {
+		readFds := &syscall.FdSet{}
+		maxfd := 0
+
+		for _, fd := range ml.fds {
+			if fd > maxfd {
+				maxfd = fd
+			}
+			addFDToFDSet(fd, readFds)
+		}
+
+		// wait for any of the sockets to be ready for accepting new connection
+		timeout := syscall.Timeval{Sec: 1, Usec: 0}
+		err := syscall.Select(maxfd+1, readFds, nil, nil, &timeout)
```

Comment:
Will this not block the OS-thread? (sorry for asking before checking)

Reply:
Yes, it will.
**Accept()** will be called inside **http.Serve** which is a blocking call. In kube-proxy we run healthz and metrics as a separate go-routine. 

https://github.com/golang/go/blob/go1.22.3/src/net/http/server.go#L3255

Reply:
also **Accept** is supposed to be a blocking call, right?

Reply:
Then I think it can't be done in this way. In go, (very) many go routines are mapped on a few OS-threads (equals the number of cores by default), and the OS-threads should not be blocked. Compare to the [resolver](https://pkg.go.dev/net#hdr-Name_Resolution). Google writes it's own resolver just to avoid blocking the OS-thread

Reply:
Would it be possible to use the normal net.Listen in some way? In go-routines and a channel for instance? And have a "super-Listen" on top

Reply:
> Then I think it can't be done in this way. In go, (very) many go routines are mapped on a few OS-threads (equals the number of cores by default), and the OS-threads should not be blocked. Compare to the [resolver](https://pkg.go.dev/net#hdr-Name_Resolution). Google writes it's own resolver just to avoid blocking the OS-thread

Maybe then there is scope for improvement in how kube-proxy handles http servers. All the servers healthz, metrics, health-check servers for services use stdlib listener which uses **Accept** which is a blocking call.

Reply:
> Would it be possible to use the normal net.Listen in some way? In go-routines and a channel for instance? And have a "super-Listen" on top

I initially planned to do that only, using go-routines to listen on every addr but then using select to mulitplex without spawning any go-routine looked a better approach to me.

Reply:
> All the servers healthz, metrics, health-check servers for services use stdlib listener which uses Accept which is a blocking call.

Are they using `syscall.Accept`?! If so, that's a bug.

Reply:
> I initially planned to do that only, using go-routines to listen on every addr but then using select to mulitplex without spawning any go-routine looked a better approach to me.

Go-routines are cheap, and I think you can get away with just one channel which makes the code simpler. Also, I think your `Listen` should take a `context` as parameter. net.Listen uses [ListenConfig](https://pkg.go.dev/net#ListenConfig) which feels like a work-around for keeping backward compatibility.

Reply:
`net.Listener.Accept()` (simplified syntax) only block the go-routine, not the OS-thread

Reply:
`l.Accept()` will block the go-routine (but not any os-thread), when a client connects it will un-block and send the conn on the channel, which will un-block multiListener.Accept. Lgtm (except that l.Accept returns (conn, err) which I assume is a typo)

Reply:
@uablrek I tried to learn more about the go-scheduler and here is my understanding:

> Then I think it can't be done in this way. In go, (very) many go routines are mapped on a few OS-threads (equals the number of cores by default),

The scheduler is not limited to `GOMAXPROCS`, it tries to keep `GOMAXPROCS` os-threads and parks any additional os-thread (puts it to sleep for reuse in future). The scheduler also has a "handoff" mechanism, when the os-thread is blocked by a syscall the scheduler unparks a thread or creates a new one if none is available and then the run-queue of the blocked os-thread is handed over to the un-parked/new thread.

> l.Accept() will block the go-routine (but not any os-thread)

l.Accept() will also block os-thread as internally it uses blocking accept(4) syscall, scheduler will unpark a thread and handoff the runqueue to the unparked thread and wait on accept(4) to complete.

https://github.com/golang/go/blob/go1.22.3/src/runtime/proc.go#L21-L31

I think calling select on all sockets will be less expensive as handoff will happen for only one os-thread compared to blocking on accept(4) on multiple sockets which will trigger multiple handoffs.

Reply:
No, a Listener.Accept() does not spawn a new OS-thread. Do an experiment:
```go
package main;
import (
	"net"
	"sync"
)
func main() {
	var wg sync.WaitGroup
	l, _ := net.Listen("tcp", "[::1]:5006")
	for i := 0; i < 100; i++ {
		wg.Add(1)
		go func() {
			_, _ = l.Accept()
		}()
	}
	wg.Wait()
}
```
Compile and run:
```
GOMAXPROCS=1 ./accept
```
Check /proc:
```
> ps awwwxxxx|grep accept
 186386 pts/5    Sl+    0:00 ./accept
 186391 pts/2    S+     0:00 grep accept
> ls /proc/186386/task/
186386/  186387/  186388/  186389/
```
OS-threads are quite resource demanding, and starting, say 10,000 OS-threads would probably crash your PC, or at least stall it. On the other hand, go is made to make go-routines cheap.

Reply:
The go "runtime" does asynchronous "accept" (and other syscall's) and wake up the appropriate go-routines. (I haven't actually read the source, but this is how it must work). Go-routines are what's called [green threads](https://en.wikipedia.org/wiki/Green_thread) back in my days.

Reply:
@uablrek I ran your code with **strace** to understand what happens under the hood and made some interesting observations:

_socket(AF_INET6, SOCK_STREAM|SOCK_CLOEXEC|SOCK_NONBLOCK, IPPROTO_IP) = 3_
**net.Listen** creates TCP socket with **SOCK_NONBLOCK**

_bind(3, {sa_family=AF_INET6, sin6_port=htons(5006), sin6_flowinfo=htonl(0), inet_pton(AF_INET6, "::1", &sin6_addr), sin6_scope_id=0}, 28) = 0_
socket binding

_listen(3, 8192)_ 
socket listen

_epoll_ctl(5, EPOLL_CTL_ADD, 3, {events=EPOLLIN|EPOLLOUT|EPOLLRDHUP|EPOLLET, data={u32=3196059650, u64=9180667291895857154}}) = 0_
This looked unexpected in the beginning, looked into the source code and found **Accept()** on **net.TCPListener** calls this to interact with [netpoller](https://morsmachine.dk/netpoller)
https://github.com/golang/go/blob/go1.22.3/src/internal/poll/fd_unix.go#L598

_accept4(3, 0xc0000d3d14, [112], SOCK_CLOEXEC|SOCK_NONBLOCK) = -1 EAGAIN (Resource temporarily unavailable)_
expected **Accept4** to fail with **EAGAIN** as socket was created using **SOCK_NONBLOCK**

_epoll_pwait(5, [], 128, 0, NULL, 0)     = 0_
This also looked unexpected, figured out from the source code that when **Accept4** fails with **EAGAIN** go waits for the underlying file descriptor to be ready using **netpoller** and that blocks the go-routine instead of os-thread.
https://github.com/golang/go/blob/go1.22.3/src/internal/poll/fd_unix.go#L611

On successful connection with a client the server flow resumes with
_accept4(3, {sa_family=AF_INET6, sin6_port=htons(52764), sin6_flowinfo=htonl(0), inet_pton(AF_INET6, "::1", 
&sin6_addr), sin6_scope_id=0}, [112 => 28], SOCK_CLOEXEC|SOCK_NONBLOCK) = 4_
This time **Accept4** is successful as the **fd** is ready for read.

And then the loop **Accept4** failing with **EAGAIN** and **EpollPwait** continues for 100 times (100 is the number of go-routines spawned in the example). 
I was expecting 100 concurrent Accept calls but that didn't happen because Accept is protected by lock. For a given socket/fd go will internally serialize all the Accept calls.
https://github.com/golang/go/blob/go1.22.3/src/internal/poll/fd_unix.go#L593-L596

TLDR;
 Calling **Accept** on **net.Listerner** doesn't block os-thread(thanks to netpoller). The current impl using **Select** will block the os-thread which will lead to performance issues.
 I will rework on the impl to use stdlib **net.Listen** in multiple go-routines instead of **Select** for multiplexing.